src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -1,53 +1,277 @@
 import { Colors } from '../../../../common/util/colors.js';
-import {
-  TypedArrayBufferView,
-  TypedArrayBufferViewConstructor,
-} from '../../../../common/util/util.js';
 import { GPUTest } from '../../../gpu_test.js';
-import { NumberType, NumberRepr } from '../../../util/conversion.js';
+import {
+  ScalarType,
+  Scalar,
+  Vector,
+  Value,
+  Type,
+  TypeVec,
+  TypeU32,
+  VectorType,
+} from '../../../util/conversion.js';
+
+/** Comparison describes the result of a Comparator function. */
+export interface Comparison {
+  matched: boolean; // True if the two values were considered a match
+  got: string; // The string representation of the 'got' value (possibly with markup)
+  expected: string; // The string representation of the 'expected' value (possibly with markup)
+}
+
+/** Comparator is a function that compares whether the provided value matches an expectation. */
+export interface Comparator {
+  (got: Value): Comparison;
+}
+
+// diff compares 'got' to 'expected, returning the Comparison information.
```

Comment:
'diff' seems like it succeeds when the values are different....

Should this be 'same' ?  or 'equalValue' or 'comparesAsEqual'?

Reply:
Define 'succeeds'? It does not throw or fail on difference, it simply returns a structure that says whether they match, along with a marked-up string representation of what was different.
So, I'm not sure I see why `diff` is not an appealing name.